### PR TITLE
Add initial ground server implementation

### PR DESCRIPTION
**Pull request by ZachGarcia42**
- Add basic react template
 - Add docker configuration for ground server
 - Update README

### CODE REVIEW

**Comment by maxslarsson** [comment]
Not to get too bogged down in the JavaScript framework ecosystem (cause it is ridiculous) but there are two frameworks that seem to be the most popular for React--those being Next.js and Vite. Next.js has a bunch of cool features and optimizations, and adds support for multiple pages (e.g. having a /prop and /software path) since React does not support that out of the box. But I am not sure that is necessary for our app since we just want one dashboard that is customizable by the user.

If we want a single page app, there are two main options--Create React App (the one you used) and Vite. Vite seems to have kind of replaced Create React App in the React ecosystem. I am not entirely sure as to why, from briefly reading around it seems like the main reasons are that Create React App is not very flexible, Vite is faster, and Vite has a large plugin ecosystem. What do you think about using Vite?

**Comment by maxslarsson** [comment]
Also, a note on your Dockerfile. Your Dockerfile seems to run ``npm run start`` which is the development version of React. For production, you "compile" the React to static HTML/JS/CSS files and serve those using a webserver like Nginx or Apache or Traefik. This also means our image does not need the entire React library + Node installed, it's faster, and we just need a simple webserver to serve the static files generated.

**Comment by cameron-goddard** [comment]
Also, just wondering - will this `ground-server/` directory replace the current `ground/` directory? Trying to visualize how the implementation will match the architecture diagram

**Comment by ZachGarcia42** [comment]
> Not to get too bogged down in the JavaScript framework ecosystem (cause it is ridiculous) but there are two frameworks that seem to be the most popular for React--those being Next.js and Vite. Next.js has a bunch of cool features and optimizations, and adds support for multiple pages (e.g. having a /prop and /software path) since React does not support that out of the box. But I am not sure that is necessary for our app since we just want one dashboard that is customizable by the user.
> 
> If we want a single page app, there are two main options--Create React App (the one you used) and Vite. Vite seems to have kind of replaced Create React App in the React ecosystem. I am not entirely sure as to why, from briefly reading around it seems like the main reasons are that Create React App is not very flexible, Vite is faster, and Vite has a large plugin ecosystem. What do you think about using Vite?

Yeah, I hadn't heard of Vite, but definitely sounds like that's what we should use. I'll update with that soon

**Comment by ZachGarcia42** [comment]
> Also, a note on your Dockerfile. Your Dockerfile seems to run `npm run start` which is the development version of React. For production, you "compile" the React to static HTML/JS/CSS files and serve those using a webserver like Nginx (the most popular one) or Apache or Traefik. This also means our image does not need the entire React library + Node installed, it's faster, and we just need a simple webserver to serve the static files generated.

Sounds good, I'll also look into that!

**Comment by ZachGarcia42** [comment]
> Also, just wondering - will this `ground-server/` directory replace the current `ground/` directory? Trying to visualize how the implementation will match the architecture diagram

Yeah, I'm thinking ground-server/ will replace ground/. Although we'll probably want to keep ground around (under a different name) as a simple gRPC client for testing/debugging

**Review by maxslarsson — COMMENTED** [review]
Sorry, didn't realize I had to submit this for it to be visible :)

**Comment by maxslarsson** [comment]
> > Also, a note on your Dockerfile. Your Dockerfile seems to run `npm run start` which is the development version of React. For production, you "compile" the React to static HTML/JS/CSS files and serve those using a webserver like Nginx (the most popular one) or Apache or Traefik. This also means our image does not need the entire React library + Node installed, it's faster, and we just need a simple webserver to serve the static files generated.
> 
> Sounds good, I'll also look into that!

I actually put a comment yesterday with an example Dockerfile that does just that--it builds the React and then serves it using Nginx. I just forgot to finalize my review, so my comment was not visible. Sorry about that!

File: ground-server/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+# Expose a port to run it on
+EXPOSE 3000
+
+# Install dependencies in package.json
+RUN npm install
+
+# Define the command that will actually start the container (in our case npm start)
+CMD ["npm", "start"]
+
```
[the remark this answers is not in the document]
Comment (ZachGarcia42) [reply ground-server/Dockerfile:20]: Sick, thanks so much!! This is great

**Comment by ZachGarcia42** [comment]
Thanks @maxslarsson so much for the expertise and dockerfile updates! Added a new commit with the vite+react template hosted with nginx. Tested with "docker run -it -d -p 80:80 zachgarcia42/ground-server-image". Would love if others could verify this on different machines. (Instructions at top of README). The image is soooo much smaller now (Woo!)

**Comment by maxslarsson** [comment]
I ran the image locally as well and it worked great! Thank you so much for putting all this together!

**Review by cameron-goddard — APPROVED** [review]

**Pull request merged**